File: windows/fortran/examples/sd_rd.f
C *****************************************************************
C                                                                 
C  sd_rd.f                                                       
C        This program creates an HDF file, sd.hdf, which        
C        contains an SDS of size 10 x 5.  It then re-opens       
C        the HDF file in Read Only mode, reads the data back     
C        in, and prints it to the screen.                        
C                                                                
C ******************************************************************

      PROGRAM SD_RD
 
      implicit none

      integer*4 sd_id, sds_id, rank, sd_index
      integer dims(2), start(2), edges(2), stride(2)
      integer i, j, istat
      integer sfstart, sfcreate, sfwdata, sfendacc, sfend
   	  integer sfrdata, sfselect
      integer*4 DFACC_CREATE, DFNT_INT16, DFACC_READ
      integer*4 DIM1, DIM0, FAIL
      parameter (DFACC_CREATE = 4, DFNT_INT16 = 22, DIM1 = 5,
     +			 DIM0 = 10, DFACC_READ = 1, FAIL = -1)
      integer*2 array_data(DIM1, DIM0)
      integer*2 new_array(DIM1, DIM0)
      integer heprnt

C     Create and open the file and initiate the SD interface.
      sd_id = sfstart('sd.hdf', DFACC_CREATE)
      if (sd_id .eq. FAIL) then
C         istat=heprnt(0)
      else
         print*
         print*,'... sd.hdf created'
      end if

C     Define the rank and dimensions of the data set to be created.
      rank = 2
      dims(1) = DIM1
      dims(2) = DIM0

C     Create the data set.
      sds_id = sfcreate(sd_id, 'data1', DFNT_INT16, rank, dims)
      if (sds_id .eq. FAIL) then
C         istat=heprnt(0)
      else
        print*,'... data1 created'
      end if

C     Fill the stored-data array with values.
      do 20 j = 1, DIM0
         do 10 i = 1, DIM1
             array_data(i, j) = i + j - 1
10    continue
20    continue

C     Define the location, pattern, and size of the data set
C     that will be written to.
      start(1) = 0
      start(2) = 0
      edges(1) = DIM1
      edges(2) = DIM0
      stride(1) = 1
      stride(2) = 1

C     Write the stored data to the "data1" data set.
      istat = sfwdata(sds_id, start, stride, edges, array_data)
      if (istat .eq. FAIL) then
C         istat=heprnt(0)
      else
         print*,'... data written out'
      end if

C     Terminate access to the array.
      istat = sfendacc(sds_id)
      if (istat .eq. FAIL) then
C         istat=heprnt(0)
      end if


C     Terminate access to the SD interface and close the file.
      istat = sfend(sd_id)
      if (istat .eq. FAIL) then
C         istat=heprnt(0)
      else
         print*,'... file closed'
      end if


C     Read data back and write to a new array 
      sd_id = sfstart('sd.hdf', DFACC_READ)
      if (sd_id .eq. FAIL) then
C         istat=heprnt(0)
      else
         print*,'... sd.hdf opened with READ access'
      end if

      sd_index = 0
      sds_id = sfselect (sd_id, sd_index) 
      if (sds_id .eq. FAIL) then
C         istat=heprnt(0)
      end if

      istat = sfrdata (sds_id, start, stride, edges, new_array) 
      if (istat .eq. FAIL) then
C         istat=heprnt(0)
      end if

      write(*,*)
      write(*,*)'Data:'
      do 40 j=1,DIM0
         write(*,*)(new_array(i,j),i=1,DIM1)
         i=i+5 
40    continue       
      write(*,*)
      
C     Terminate access to the array.
      istat = sfendacc(sds_id)
      if (istat .eq. FAIL) then
C         istat=heprnt(0)
      end if

C     Terminate access to the SD interface and close the file.
      istat = sfend(sd_id)
      if (istat .eq. FAIL) then
C         istat=heprnt(0)
      else
         print*,'... file closed'
         print*
      end if

      end

File: windows/examples/cexamples/sd_rd.c
/*******************************************************************/
/*                                                                 */
/*  sd_rd.c                                                        */
/*        This program creates an HDF file, sd.hdf, which          */
/*        contains an SDS of size 10 x 5.  It then re-opens        */
/*        the HDF file in Read Only mode, reads the data back      */
/*        in, and prints it to the screen.                         */
/*                                                                 */
/*******************************************************************/

#include "hdf.h"

#define DIM1 5
#define DIM0 10

main( ) 
{
	int32 sd_id, sds_id, istat, sd_index;
	int32 dims[2], start[2], edges[2], rank;
	int16 array_data[DIM0][DIM1];
	int16 new_array[DIM0][DIM1]; 
	intn i, j;

	/* Create and open the file and initiate the SD interface. */
	sd_id = SDstart("sd.hdf", DFACC_CREATE);
	if (sd_id != FAIL)
		printf ("\n... sd.hdf created\n");

	/* Define the rank and dimensions of the data set to be created. */
	rank = 2;
	dims[0] = DIM0;
	dims[1] = DIM1;

	/* Create the array data set. */
	sds_id = SDcreate(sd_id, "data1", DFNT_INT16, rank, dims);
	if (sds_id != FAIL)
		printf ("... data1 created\n");


	/* Fill the stored-data array with values. */
	for (j = 0; j < DIM0; j++) {
		for (i = 0; i < DIM1; i++)
			array_data[j][i] = (i + j) + 1;
 	}

	/* Define the location, pattern, and size of the data set */
	for (i = 0; i < rank; i++) {
		start[i] = 0;
		edges[i] = dims[i];
	}

	istat = SDwritedata(sds_id, start, NULL, edges, (VOIDP)array_data); 
	if (istat != FAIL)
		printf ("... data written out\n");


	/* Terminate access to the array. */
	istat = SDendaccess(sds_id);

	/* Terminate access to the SD interface and close the file. */
	istat = SDend(sd_id);
	if (istat != FAIL)
		printf("... file closed\n");

	/* Read data back and write to a new array */

	sd_id = SDstart("sd.hdf", DFACC_READ);
	if (sd_id != FAIL)
		printf ("... sd.hdf opened with READ access\n");

	sd_index = 0;
	sds_id = SDselect (sd_id, sd_index); 

	istat = SDreaddata (sds_id, start, NULL, edges, (VOIDP) new_array);

	printf ("\nData: \n");
	for (j = 0; j < DIM0; j++) 
	{
		for (i = 0; i < DIM1; i++)
			printf ("  %i ", new_array[j][i]);
		printf ("\n");
	}
	printf ("\n");

	/* Terminate access to the array. */
	istat = SDendaccess(sds_id);

	/* Terminate access to the SD interface and close the file. */
	istat = SDend(sd_id);
	if (istat != FAIL)
		printf("... file closed\n\n");

}


